e2e/ruler_test.go

```diff
@@ -25,174 +25,212 @@ import (
+	ctx := context.Background()
+	kubeClient, opClient, err := newKubeClients()
 	if err != nil {
-		t.Fatal(err)
+		t.Fatalf("error instantiating clients. err: %s", err)
 	}
 
-	t.Run("rule evaluator secrets", tctx.subtest(func(ctx context.Context, t *OperatorContext) {
-		testRuleEvaluatorSecrets(ctx, t, cert, key)
-	}))
-	t.Run("rule evaluator config", tctx.subtest(func(ctx context.Context, t *OperatorContext) {
-		testRuleEvaluatorConfig(ctx, t, cert, key)
-	}))
-	t.Run("rule generation", tctx.subtest(testRulesGeneration))
-	t.Run("rule evaluator deploy", tctx.subtest(testRuleEvaluatorDeployment))
-
+	t.Run("rule-evaluator-deployed", testRuleEvaluatorDeployed(ctx, t, kubeClient))
+	t.Run("rule-evaluator-configured", testRuleEvaluatorConfigured(ctx, t, kubeClient, opClient))
+	// TODO(pintohutch): testing the generated secrets and config can be
+	// brittle as the checks need to be precise and could break if mechanics or
+	// formatting changes in the future.
+	// Ideally this is replaced by a true e2e test that deploys a custom
+	// secured alertmanager and successfully sends alerts to it.
+	t.Run("rule-evaluator-secrets", testRuleEvaluatorSecrets(ctx, t, kubeClient, opClient))
+	t.Run("rule-evaluator-configuration", testRuleEvaluatorConfiguration(ctx, t, kubeClient))
+
+	t.Run("rules-create", testCreateRules(ctx, t, opClient))
 	if !skipGCM {
-		t.Log("Waiting rule results to become readable")
-		t.Run("check rule metrics", tctx.subtest(testValidateRuleEvaluationMetrics))
+		t.Run("rules-gcm", testValidateRuleEvaluationMetrics(ctx, t))
+	}
+}
+
+func testRuleEvaluatorDeployed(ctx context.Context, t *testing.T, kubeClient kubernetes.Interface) func(*testing.T) {
+	return func(t *testing.T) {
+		t.Log("checking rule-evaluator is running")
+
+		err := wait.PollUntilContextCancel(ctx, pollDuration, false, func(ctx context.Context) (bool, error) {
+			deploy, err := kubeClient.AppsV1().Deployments(operator.DefaultOperatorNamespace).Get(ctx, operator.NameRuleEvaluator, metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			} else if err != nil {
+				return false, fmt.Errorf("getting rule-evaluator Deployment failed: %w", err)
+			}
+			if *deploy.Spec.Replicas != deploy.Status.ReadyReplicas {
+				return false, nil
+			}
+
+			// Assert we have the expected annotations.
+			wantedAnnotations := map[string]string{
+				"components.gke.io/component-name":               "managed_prometheus",
+				"cluster-autoscaler.kubernetes.io/safe-to-evict": "true",
+			}
+			if diff := cmp.Diff(wantedAnnotations, deploy.Spec.Template.Annotations); diff != "" {
+				return false, fmt.Errorf("unexpected annotations (-want, +got): %s", diff)
+			}
+			return true, nil
+		})
+		if err != nil {
+			t.Fatalf("failed waiting for generated rule-evaluator deployment: %s", err)
+		}
+	}
+}
+
+func testRuleEvaluatorConfigured(ctx context.Context, t *testing.T, kubeClient kubernetes.Interface, opClient versioned.Interface) func(*testing.T) {
```

Comment:
Hm, the name `testRuleEvaluatorConfigured` and `testRuleEvaluatorConfiguration` are a bit too close, it's hard to understand from function name what's for what. This also indicated the problem of the bit duplicated purpose of those. Can we have one where we ensure OperatorConfig has a custom change and we check rule-eval Prometheus like config?

Reply:
The same applies to all other tests in other PRs. A bit confusing 🤔

Reply:
Agreed. Changed the name along with the new pattern from collector_test.go to make this more clear.